sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala

```diff
@@ -92,46 +89,53 @@ object GenerateUnsafeProjection extends CodeGenerator[Seq[Expression], UnsafePro
       s"$rowWriter.reset();"
     }
 
-    val writeFields = inputs.zip(inputTypes).zipWithIndex.map {
-      case ((input, dataType), index) =>
+    val writeFields = inputs.zip(inputsTypeAndNullable).zipWithIndex.map {
+      case ((input, (dataType, nullable)), index) =>
         val dt = dataType match {
           case udt: UserDefinedType[_] => udt.sqlType
           case other => other
         }
         val tmpCursor = ctx.freshName("tmpCursor")
 
         val setNull = dt match {
-          case t: DecimalType if t.precision > Decimal.MAX_LONG_DIGITS =>
-            // Can't call setNullAt() for DecimalType with precision larger than 18.
-            s"$rowWriter.write($index, (Decimal) null, ${t.precision}, ${t.scale});"
+          case dt if UnsafeRow.isMutable(dt) && !UnsafeRow.isFixedLength(dt) =>
+            def varLenDataSize(s: DataType): Int = s match {
+              case t: DecimalType if t.precision > Decimal.MAX_LONG_DIGITS => 16
```

Comment:
If `!UnsafeRow.isFixedLength(dt)` and `dt` is `DecimalType`, I think it already implicitly means `t.precision > Decimal.MAX_LONG_DIGITS`?